src/coprocessor/gadgets.rs

```diff
@@ -124,22 +134,364 @@ pub fn construct_list<F: LurkField, CS: ConstraintSystem<F>>(
+}
+
+/// Deconstructs an `AllocatedPtr` that's assumed to encode a cons-list. It
+/// requires prior knowledge of the total number `n` of elements in the list,
+/// including a possibly terminating `nil`.
+pub fn deconstruct_list<F: LurkField, CS: ConstraintSystem<F>>(
+    cs: &mut CS,
+    g: &GlobalAllocator<F>,
+    not_dummy: &Boolean,
+    store: &Store<F>,
+    list: &AllocatedPtr<F>,
+    n: usize,
+) -> Result<Vec<AllocatedPtr<F>>, SynthesisError> {
+    assert!(n > 0, "list must be at least nil");
+    // get the `ZPtr` vector
+    let z_ptrs = if not_dummy.get_value() == Some(true) {
+        let (mut ptrs, last) = store
+            .fetch_list(&get_ptr(list, store)?)
+            .ok_or_else(|| SynthesisError::Unsatisfiable)?;
+        ptrs.push(last.unwrap_or_else(|| store.intern_nil()));
+        assert_eq!(ptrs.len(), n);
+        ptrs.iter().map(|ptr| store.hash_ptr(ptr)).collect()
+    } else {
+        vec![ZPtr::dummy(); n]
+    };
+
+    // allocate the pointers to return
+    let mut a_ptrs = Vec::with_capacity(z_ptrs.len());
+    // the allocated pointers are reversed for now
+    for (i, z_ptr) in z_ptrs.into_iter().rev().enumerate() {
+        a_ptrs.push(AllocatedPtr::alloc(
+            &mut cs.namespace(|| format!("ptr {i}")),
+            || Ok(z_ptr),
+        )?)
+    }
+
+    if n == 1 {
+        // list is nil
+        let a_nil = g
+            .get_allocated_ptr_from_ptr(&store.intern_nil(), store)
+            .expect("nil pointer not allocated");
+        a_ptrs[0].implies_ptr_equal(&mut cs.namespace(|| "nil ptr equality"), not_dummy, &a_nil);
+    } else {
+        // enforce the final hash as the accumulated cons-list hashing
+        let init = &a_ptrs[0];
+        let cons_tag = g.get_tag(&ExprTag::Cons).expect("Cons tag not allocated");
+        let mut tag = init.tag();
+        let mut hash = init.hash().clone();
+        for (i, a_ptr) in a_ptrs[1..].iter().enumerate() {
+            hash = hash_poseidon(
+                &mut cs.namespace(|| format!("hash {i}")),
+                vec![a_ptr.tag().clone(), a_ptr.hash().clone(), tag.clone(), hash],
+                store.poseidon_cache.constants.c4(),
+            )?;
+            tag = cons_tag;
+        }
+        implies_equal(
+            &mut cs.namespace(|| "hash equality"),
+            not_dummy,
+            list.hash(),
+            &hash,
+        );
+        // recover the correct order
+        a_ptrs.reverse();
+    }
+    Ok(a_ptrs)
+}
+
+/// Deconstructs an `AllocatedPtr` with `car_cdr` semantics
+pub fn car_cdr<F: LurkField, CS: ConstraintSystem<F>>(
+    cs: &mut CS,
+    g: &GlobalAllocator<F>,
+    store: &Store<F>,
+    not_dummy: &Boolean,
+    maybe_cons: &AllocatedPtr<F>,
+) -> Result<(AllocatedPtr<F>, AllocatedPtr<F>), SynthesisError> {
+    let (car, cdr) = if not_dummy.get_value() == Some(true) {
+        let (car, cdr) = store
+            .car_cdr(&get_ptr(maybe_cons, store)?)
+            .expect("invalid Ptr");
```

Comment:
Please document that this function will panic if `maybe_cons` is missing or is not a cons.